docs/docs/how_to/how-to-oracles.md

```diff
@@ -0,0 +1,270 @@
+
+```js
+server.addMethod("getSquared", async (params) => {
+  const values = params[0].Array.map(({ inner }) => {
+    return { inner: `${inner * inner}` };
+  });
+  return { values: [{ Array: values }] };
+});
+```
+
+:::tip
+
+Brillig expects an object with an array of values. Each value is an object declaring to be `Single` or `Array` and returning a `inner` property *as a string*. For example:
+
+```json
+{ "values": [{ "Array": [{ "inner": "1" }, { "inner": "2"}]}]}
+{ "values": [{ "Single": { "inner": "1" }}]}
+{ "values": [{ "Single": { "inner": "1" }}, { "Array": [{ "inner": "1", { "inner": "2" }}]}]}
+```
+
+If you're using Typescript, the following types may be helpful in understanding the expected return value and making sure they're easy to follow:
+
+```js
+interface Value {
+  inner: string,
+}
+
+interface SingleForeignCallParam {
+  Single: Value,
+}
+
+interface ArrayForeignCallParam {
+  Array: Value[],
+}
+
+type ForeignCallParam = SingleForeignCallParam | ArrayForeignCallParam;
+
+interface ForeignCallResult {
+  values: ForeignCallParam[],
+}
+```
+
+:::
+
+## Step 3 - Usage with Nargo
+
+Using the [`nargo` CLI tool](../getting_started/installation/index.md), you can use oracles in the `nargo test`, `nargo execute` and `nargo prove` commands by passing a value to `--oracle-resolver`. For example:
+
+```bash
+nargo test --oracle-resolver http://localhost:5555
+```
+
+This tells `nargo` to use your RPC Server URL whenever it finds an oracle decorator.
+
+## Step 4 - Usage with NoirJS
+
+In a JS environment, an RPC server is not strictly necessary, as you may want to resolve your oracles without needing any JSON call at all. NoirJS simply expects that you pass a callback function when you generate proofs, and that callback function can be anything.
+
+For example, if your circuit expects the host machine to provide CPU pseudo-randomness, you could simply pass it as the `foreignCallHandler`. You don't strictly need to create an RPC server to serve pseudo-randomness, as you may as well get it directly in your app:
+
+```js
+const foreignCallHandler = (name, inputs) => crypto.randomBytes(16) // etc
+
+await noir.generateFinalProof(inputs, foreignCallHandler)
+```
+
+As one can see, in NoirJS, the [`foreignCallHandler`](../reference/NoirJS/noir_js/type-aliases/ForeignCallHandler.md) function simply means "a callback function that returns a value of type [`ForeignCallOutput`](../reference/NoirJS/noir_js/type-aliases/ForeignCallOutput.md). It doesn't have to be an RPC call like in the case for Nargo.
+
+:::tip
+
+Does this mean you don't have to write an RPC server like in [Step #2](#step-2---write-an-rpc-server)?
+
+You don't technically have to, but then how would you run `nargo test` or `nargo prove`? To use both `Nargo` and `NoirJS` in your development flow, you will most certainly have to write a JSON RPC server anyway.
```

Comment:
less :slightly_smiling_face: 
```suggestion
You don't technically have to, but then how would you run `nargo test` or `nargo prove`? To use both `Nargo` and `NoirJS` in your development flow, you will have to write a JSON RPC server.
```

Reply:
Lessssss 😈

(we should put that as #2 rule of devrel - #1 will always be "tag a DevRel")